lib/spack/spack/solver/concretize.lp

```diff
@@ -291,14 +291,28 @@ pkg_fact(Package, version_declared(Version)) :- pkg_fact(Package, version_declar
 % is not precisely one version chosen. Error facts are heavily optimized
 % against to ensure they cannot be inferred when a non-error solution is
 % possible
-{ attr("version", node(ID, Package), Version) : pkg_fact(Package, version_declared(Version)) }
+{ choose_version(node(ID, Package), Version) : pkg_fact(Package, version_declared(Version)) }
  :- attr("node", node(ID, Package)).
 
+#defined compiler_package/1.
+
+attr("version", node(ID, Package), Version) :- choose_version(node(ID, Package), Version).
+
 % A virtual package may or may not have a version, but never has more than one
 error(100, "Cannot select a single version for virtual '{0}'", Virtual)
   :- attr("virtual_node", node(ID, Virtual)),
      2 { attr("version", node(ID, Virtual), Version) }.
 
+% This exists only to improve error messages: normally selecting a
+% version with a choice rule is penalized. This avoids the penalty
+% when the version comes from an activated requirement.
+% Oddly, does not seem to work if I relocate it to error_messages.lp
+{ attr("version", node(ID, Package), Version) } :-
+  requirement_group_member(ConditionID, Package, RequirementID),
+  activate_requirement(node(ID, Package), RequirementID),
+  pkg_fact(Package, condition_effect(ConditionID, EffectID)),
+  imposed_constraint(EffectID, "node_version_satisfies", Package, Version).
```

Comment:
Some timing with requirements on would have been nice in this PR, so to know if there is any slow-down / speed-up to be expected.

Reply:
> How do I do that? (i.e. how do I produce the before/after stats that you add to some PRs)?

We have a Spack extension here https://github.com/spack/spack-benchmark In the README you'll find the commands.